parser/src/aspects/loop.rs

```diff
@@ -42,20 +49,239 @@ impl<'a> LoopAspect<'a> for Parser<'a> {
+
+        Ok(For { kind, body })
+    }
+}
+
+impl<'a> Parser<'a> {
+    /// Parses the for kind, either a range for or a conditional for.
+    fn parse_for_kind(&mut self) -> ParseResult<ForKind<'a>> {
+        let current = self.cursor.peek();
+        let start_pos = self.cursor.relative_pos(&current).start;
+        match current.token_type {
+            TokenType::Identifier => {
+                let range_for = self.parse_range_for()?;
+                Ok(ForKind::Range(range_for))
+            }
+            TokenType::RoundedLeftBracket => {
+                let conditional_for = self.parse_conditional_for()?;
+                Ok(ForKind::Conditional(conditional_for))
+            }
+
+            // Those tokens are not valid here, but we can try to parse them and give a better error
+            // message.
+            TokenType::Dollar => {
+                self.cursor.next_opt();
+                if self.parse_range_for().is_ok() {
+                    let end_pos = self.cursor.relative_pos(&self.cursor.peek()).end;
+                    let slice = &self.source.source[start_pos + 1..end_pos];
+                    return self.expected_with(
+                        "Receiver variables do not start with '$'.",
+                        current,
+                        ParseErrorKind::UnexpectedInContext(format!(
+                            "Consider removing the '$' prefix: for {slice}"
+                        )),
+                    );
+                }
+                self.expected("Expected identifier", ParseErrorKind::Unexpected)
+            }
+            TokenType::In => self.expected(
+                "Expected variable name before 'in'",
+                ParseErrorKind::Unexpected,
+            ),
+            _ => self.expected("Expected identifier or '(('", ParseErrorKind::Unexpected),
+        }
+    }
+
+    /// Parses a for loop with range, with a receiver and an iterable.
+    fn parse_range_for(&mut self) -> ParseResult<RangeFor<'a>> {
+        let receiver = self.cursor.force(
+            of_type(TokenType::Identifier),
+            "Excepted a variable identifier",
+        )?;
+        self.cursor.advance(blanks());
+        self.cursor.force(
+            of_type(TokenType::In),
+            "expected 'in' after receiver in range for",
+        )?;
+        self.cursor.advance(blanks());
+        let iterable = self.parse_iterable()?;
+
+        Ok(RangeFor {
+            receiver: receiver.value,
+            iterable,
+        })
+    }
+
+    /// Parses a "traditional" conditional for, with a initializer, a condition and an increment.
+    fn parse_conditional_for(&mut self) -> ParseResult<ConditionalFor<'a>> {
+        let start = self.cursor.force(
+            repeat_nm(2, 2, of_type(TokenType::RoundedLeftBracket)),
+            "expected '((' at start of conditional for",
+        )?;
+        for _ in 0..2 {
+            self.delimiter_stack.push_back(start.clone());
+        }
```

Comment:
why push the _same_ delimiter two times in the stack as a closing delimiter such as `)<x>)` is invalid.